tools/convert_pkl_to_pb.py

```diff
@@ -124,10 +125,41 @@ def unscope_name(name):
 
 
 def reset_names(names):
-    for i in range(0, len(names)):
+    for i in range(len(names)):
         names[i] = unscope_name(names[i])
 
 
+def convert_collect_and_distribute(
+    op, blobs,
+    roi_canonical_scale,
+    roi_canonical_level,
+    roi_max_level,
+    roi_min_level,
+    rpn_max_level,
+    rpn_min_level,
+    rpn_post_nms_topN,
+):
+    print('Converting CollectAndDistributeFpnRpnProposals'
+          ' Python -> C++:\n{}'.format(op))
+    assert op.name.startswith('CollectAndDistributeFpnRpnProposalsOp'), \
+        'Not valid CollectAndDistributeFpnRpnProposalsOp'
+
+    inputs = [x for x in op.input]
+    ret = core.CreateOperator(
+        'CollectAndDistributeFpnRpnProposals',
+        inputs,
```

Comment:
Why not use <code>list(op.input)</code> ? It seems to have the exact same output.

Reply:
Yes, it is from @orionr 's code. I have no idea whether it is better for me to change it.

Reply:
Hi @daquexian,

Thanks for writing this. While running it to convert pkl model(FPN), I get an error at     verify_model(args, [net, init_net], args.test_img)

Something like as shown below:

AssertionError:
Arrays are not almost equal to 3 decimals
result_boxes and result_boxes not matched. Max diff: 5.0
(mismatch 0.606060606061%)
 x: array([[3.329e+02, 1.799e+01, 5.250e+02, 3.556e+02, 9.995e-01],
       [0.000e+00, 2.073e+01, 2.308e+02, 3.685e+02, 9.926e-01],
       [2.459e+01, 3.320e+01, 1.330e+02, 1.092e+02, 9.909e-01],...
 y: array([[3.329e+02, 1.799e+01, 5.250e+02, 3.556e+02, 9.995e-01],
       [0.000e+00, 2.073e+01, 2.308e+02, 3.685e+02, 9.926e-01],
       [2.459e+01, 3.320e+01, 1.330e+02, 1.092e+02, 9.909e-01],...

I get this error when I use a test_img, otherwise model gets written in pb format.

How important it is to get correct boxes on test image??

Reply:
@daquexian Sure. I can share the model, config file and test image. How would you like me to share it?

Reply:
@daquexian : I was able to convert pkl model(with fpn) to protobufs. However, when I try to run this net, I get following error:

RuntimeError: [enforce fail at utility_ops.h:275] . Check failed: output->dims() == Input(i).dims().Description: Input #1, input dimension:1 256 38 60 should match output dimension: 1 256 38 59 Error from operator: 
input: "fpn_inner_res4_5_sum_lateral" input: "fpn_inner_res4_5_sum_topdown" output: "fpn_inner_res4_5_sum" name: "" type: "Sum" device_option { device_type: 1 cuda_gpu_id: 0 }


The error is typical of models with fpn layers. I never encountered this kind of issue before. Any pointers to what might be causing this would be of great help.

Reply:
@rohitbhio I think it is caused by a wrong `FPN.COARSEST_STRIDE`. Are you using the latest code? Is your `FPN.COARSEST_STRIDE` in yaml right?

Reply:
@daquexian While training model with FPN, I didn't try overwriting `FPN.COARSEST_STRIDE` in my config file. Meaning I used default `FPN.COARSEST_STRIDE` = 32. What is the recommended value of COARSEST_STRIDE btw?

Reply:
@daquexian : pasting my config file below

MODEL:
  TYPE: generalized_rcnn
  CONV_BODY: FPN.add_fpn_ResNet50_conv5_body
  NUM_CLASSES: 3 #Including Background
  FASTER_RCNN: True
NUM_GPUS: 1
SOLVER:
  WEIGHT_DECAY: 0.0001
  LR_POLICY: steps_with_decay
  BASE_LR: 0.0025
  GAMMA: 0.1
  MAX_ITER: 40000
  STEPS: [0, 10000, 13000, 25000,32000]
FPN:
  FPN_ON: True
  MULTILEVEL_ROIS: True
  MULTILEVEL_RPN: True
FAST_RCNN:
  ROI_BOX_HEAD: fast_rcnn_heads.add_roi_2mlp_head
  ROI_XFORM_METHOD: RoIAlign
  ROI_XFORM_RESOLUTION: 7
  ROI_XFORM_SAMPLING_RATIO: 2
TRAIN:
  WEIGHTS: https://s3-us-west-2.amazonaws.com/detectron/ImageNetPretrained/MSRA/R-50.pkl
  #WEIGHTS: /rcode/detectron-output/train/person_hardhat_det_train/generalized_rcnn/model_final.pkl
  DATASETS: ('person_hardhat_det_train',)
  SCALES: (500,)
  MAX_SIZE: 833
  BATCH_SIZE_PER_IM: 256
  RPN_PRE_NMS_TOP_N: 2000  # Per FPN level
TEST:
  WEIGHTS: /rcode/Workspace/Rohit/resnetfpn/train/person_hardhat_det_train/generalized_rcnn/model_final.pkl
  FORCE_JSON_DATASET_EVAL: True
  DATASETS: ('person_hardhat_det_val',)
  SCALE: 500
  MAX_SIZE: 833
  NMS: 0.5
  RPN_PRE_NMS_TOP_N: 1000  # Per FPN level
  RPN_POST_NMS_TOP_N: 1000
OUTPUT_DIR: .

Reply:
@rohitbhio 32 is right for your case. Are you using the latest code?

Reply:
@daquexian I trained my model with 2 months old codebase. I am training the same model using latest code. I will let you know the outcome.

If I resize my image to (640,640), will it not throw an error since my net expects an input dimension of (500,833)? Anyways, I will try this.

Reply:
Hi @daquexian. I have trained new model using latest code. I still encounter same error reported earlier

Reply:
@rohitbhio Thanks. Could you please tell me the shape of `blob`(it's in line 490 of `convert_pkl_to_pb.py`)? Could you please try to forcely resize `im` to (640, 640) by modifying line 486 in `convert_pkl_to_pb.py`(this line should be `im=cv2.resize(.....)`)? What about forcely resize to (672, 672)? Then what about forcely resize `blob` instead of `im`?

Reply:
@rohitbhio Are you using the latest code of this PR? Is `blob_utils.im_list_to_blob([im])` in `convert_pkl_to_pb.py` in your code? What's the shape of `blob` if you don't forcely resize it, could you please print it and tell me?

Reply:
@daquexian : Yes, there is a method `blob_utils.im_list_to_blob([im])` in `convert_pkl_to_pb.py`. The size of original image is `(377, 590, 3)`.  

The size of the blob while data prep becomes `(1, 3, 500, 782)`. Also, content of `im_info` is `array([[500. , 782. , 1.32626]], dtype=float32)`

Reply:
@rohitbhio The error you encountered is caused by input size which is not a multiple of 32. When the size is not a multiple of 32, for example, the width is 36, after a certain layer, a feature map whose width=36/2/2=9 is produced, then after a conv layer whose stride=2 and a upsample2x layer, a feature map whose width=8 instead of 9 is produced, then an error occurs when summing these two feature maps whose widths are different with each other. `blob_utils.im_list_to_blob` pads the input so that its height and width will be a multiple of 32. You can try to print some messages to check whether `blob_utils.im_list_to_blob` works correctly.

Reply:
@daquexian: Thanks for the explanation. Your observation regarding `blob_utils.im_list_to_blob` method is correct. For some reason, the method was not adjusting image blob dimensions according to `COARSEST_STRIDE`.  Now, the size of the blob while data prep becomes (1, 3, 512, 800) and I am able to make detections. Wow. Thanks.

Reply:
@rohitbhio Glad to see it :) Could you please tell me why the method doesn't adjust the dimensions? Is there anything wrong in my PR?

Reply:
@daquexian Your PR is awesome. I was trying to load these protobuf models without loading its corresponding config yaml file. Now, `blob_utils.im_list_to_blob` method ended up using `cfg.FPN.FPN_ON` and `cfg.FPN.COARSEST_STRIDE` from default config file(config.py).  In default config file, `cfg.FPN.FPN_ON` is set to `False`.  For my purpose, I made a copy of `im_list_to_blob` method from blob.py and modified it so that I could explicitly pass `FPN_ON` and `COARSEST_STRIDE` as input parameters so that this method doesn't take global values.